File: src/leetcode/2026/04/Solution783.java
class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;

    TreeNode() {
    }

    TreeNode(int val) {
        this.val = val;
    }

    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

public class Solution783 {
    public int minDiffInBST(TreeNode root) {
        int minDiff = Integer.MAX_VALUE;
        TreeNode prev = null;

        while (root != null) {
            if (root.left == null) {
                if (prev != null) {
                    minDiff = Math.min(minDiff, root.val - prev.val);
                }
                prev = root;
                root = root.right;
            } else {
                TreeNode predecessor = root.left;
                while (predecessor.right != null && predecessor.right != root) {
                    predecessor = predecessor.right;
                }

                if (predecessor.right == null) {
                    predecessor.right = root;
                    root = root.left;
                } else {
                    predecessor.right = null;
                    if (prev != null) {
                        minDiff = Math.min(minDiff, root.val - prev.val);
                    }
                    prev = root;
                    root = root.right;
                }
            }
        }

        return minDiff;
    }

    public static void main(String[] args) {
        Solution783 solution = new Solution783();

        TreeNode root = new TreeNode(4);
        root.left = new TreeNode(2);
        root.right = new TreeNode(6);
        root.left.left = new TreeNode(1);
        root.left.right = new TreeNode(3);

        int output = solution.minDiffInBST(root);
        System.out.println("Output: " + output);

        root = new TreeNode(1);
        root.right = new TreeNode(3);

        output = solution.minDiffInBST(root);
        System.out.println("Output: " + output);
    }
}

File: src/leetcode/2026/04/Solution783.cpp
#include <climits>
#include <algorithm>
using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution783 {
public:
    int minDiffInBST(TreeNode* root) {
        int min_diff = INT_MAX;
        TreeNode* prev = nullptr;
        dfs(root, prev, min_diff);
        return min_diff;
    }

private:
    void dfs(TreeNode* node, TreeNode*& prev, int& min_diff) {
        if (!node) return;
        dfs(node->left, prev, min_diff);
        if (prev) {
            min_diff = min(min_diff, node->val - prev->val);
        }
        prev = node;
        dfs(node->right, prev, min_diff);
    }
};

int main() {
    Solution783 solution;
    TreeNode* root = new TreeNode(4);
    root->left = new TreeNode(2);
    root->right = new TreeNode(6);
    root->left->left = new TreeNode(1);
    root->left->right = new TreeNode(3);
    
    int result = solution.minDiffInBST(root);
    printf("Minimum difference in BST: %d\n", result); // Output: 1

    // Clean up memory
    delete root->left->left;
    delete root->left->right;
    delete root->left;
    delete root->right;
    delete root;

    return 0;
}